crates/wasi-common/tokio/src/lib.rs

```diff
@@ -0,0 +1,109 @@
+            random_ctx(),
+            clocks_ctx(),
+            sched_ctx(),
+            Rc::new(RefCell::new(Table::new())),
+        ))
+    }
+    pub fn env(self, var: &str, value: &str) -> Result<Self, wasi_common::StringArrayError> {
+        let s = self.0.env(var, value)?;
+        Ok(WasiCtxBuilder(s))
+    }
+    pub fn envs(self, env: &[(String, String)]) -> Result<Self, wasi_common::StringArrayError> {
+        let mut s = self;
+        for (k, v) in env {
+            s = s.env(k, v)?;
+        }
+        Ok(s)
+    }
+    pub fn inherit_env(self) -> Result<Self, wasi_common::StringArrayError> {
+        let mut s = self.0;
+        for (key, value) in std::env::vars() {
+            s = s.env(&key, &value)?;
+        }
+        Ok(WasiCtxBuilder(s))
+    }
+    pub fn arg(self, arg: &str) -> Result<Self, wasi_common::StringArrayError> {
+        let s = self.0.arg(arg)?;
+        Ok(WasiCtxBuilder(s))
+    }
+    pub fn args(self, arg: &[String]) -> Result<Self, wasi_common::StringArrayError> {
+        let mut s = self;
+        for a in arg {
+            s = s.arg(&a)?;
+        }
+        Ok(s)
+    }
+    pub fn inherit_args(self) -> Result<Self, wasi_common::StringArrayError> {
+        let mut s = self.0;
+        for arg in std::env::args() {
+            s = s.arg(&arg)?;
+        }
+        Ok(WasiCtxBuilder(s))
+    }
+    pub fn stdin(self, f: Box<dyn wasi_common::WasiFile>) -> Self {
+        WasiCtxBuilder(self.0.stdin(f))
+    }
+    pub fn stdout(self, f: Box<dyn wasi_common::WasiFile>) -> Self {
+        WasiCtxBuilder(self.0.stdout(f))
+    }
+    pub fn stderr(self, f: Box<dyn wasi_common::WasiFile>) -> Self {
+        WasiCtxBuilder(self.0.stderr(f))
+    }
+    pub fn inherit_stdin(self) -> Self {
+        self.stdin(Box::new(crate::stdio::stdin()))
+    }
+    pub fn inherit_stdout(self) -> Self {
+        self.stdout(Box::new(crate::stdio::stdout()))
+    }
+    pub fn inherit_stderr(self) -> Self {
+        self.stderr(Box::new(crate::stdio::stderr()))
+    }
+    pub fn inherit_stdio(self) -> Self {
+        self.inherit_stdin().inherit_stdout().inherit_stderr()
+    }
+    pub fn preopened_dir(
+        self,
+        dir: cap_std::fs::Dir,
+        guest_path: impl AsRef<Path>,
+    ) -> Result<Self, wasi_common::Error> {
+        let dir = Box::new(Dir::from_cap_std(dir));
+        Ok(WasiCtxBuilder(self.0.preopened_dir(dir, guest_path)?))
+    }
+    pub fn build(self) -> Result<WasiCtx, wasi_common::Error> {
+        self.0.build()
+    }
+}
+
+// This function takes the "async" code which is in fact blocking
+// but always returns Poll::Ready, and executes it in a dummy executor
+// on a blocking thread in the tokio thread pool.
+pub(crate) fn asyncify<'a, F, Fut, T>(f: F) -> Result<T, Error>
+where
+    F: FnOnce() -> Fut + Send + 'a,
+    Fut: Future<Output = Result<T, Error>>,
+    T: Send + 'static,
+{
+    tokio::task::block_in_place(move || unsafe { wiggle::run_in_dummy_executor(f()) })
+}
```

Comment:
I think I'm getting a bit lost in my head thinking about this. This feels pretty backwards where we're "asyncifying" already async code, and I'm not sure that I have enough context for untangling why the backwards feeling is actually ok here.

The idea here is that wasi-common requires everything to be async, but then the "sync" versions just do all the blocking work in their async impls. Then when tokio builds on top of the sync versions again it just rewraps everything in `block_in_place`? Basically the `async` in cap-std-sync is a lie?

I think this may want a more clear comment somewhere if possible?

Reply:
Agreed this needs a comment. The `async` in cap-std-sync is indeed a lie.